### PR TITLE
[ASTMangler] Do not crash when mangling a retroactive conformance

### PR DESCRIPTION
**Pull request by theblixguy**
We have a crash when trying to get the conformance access path in `appendConcreteProtocolConformance` because `CurGenericSignature` is null. 

I think the reason why this is happening is because we're calling `appendType` before calling `appendGenericSignature` (or some other method which also sets `CurGenericSignature`), which means in some cases we don't have a `CurGenericSignature` to grab the conformance path from when mangling a retroactive conformance.

For example, the stack trace for the failure case looks like this:

```
-> IRGenMangler::mangleTypeForReflection
  -> IRGenMangler::withSymbolicReferences
    -> ASTMangler::appendType
      -> ASTMangler::appendRetroactiveConformances
        -> ASTMangler::appendConcreteProtocolConformance
```

Another one looks like this:

```
-> getMangledName
  -> ASTMangler::mangleTypeForDebugger
    -> ASTMangler::appendType
      -> ASTMangler::appendRetroactiveConformances
        -> ASTMangler::appendConcreteProtocolConformance
```

In both these paths, when `appendConcreteProtocolConformance ` is called, we don't have a `CurGenericSignature`. For example - in the second scenario, `mangleTypeForDebugger` is called with a null `DeclContext` so we never call `bindGenericParameters` which would then go on to set `CurGenericSignature`.

To fix this, I have made `appendConcreteProtocolConformance` more defensive. I am not sure if this is the right thing to do, but it makes the failing cases compile & work.

Resolves [SR-11521](https://bugs.swift.org/browse/SR-11521)
Resolves [SR-11561](https://bugs.swift.org/browse/SR-11561)
Resolves [SR-11573](https://bugs.swift.org/browse/SR-11573)
Resolves rdar://problem/55705702
Resolves rdar://problem/55987394
Resolves feedback://7351428

### CODE REVIEW

**Comment by theblixguy** [comment]
ping @jckarter @DougGregor

File: lib/AST/ASTMangler.cpp [hunk 1]
```diff
@@ -2649,6 +2649,10 @@ void ASTMangler::appendDependentProtocolConformance(
 void ASTMangler::appendConcreteProtocolConformance(
                                       const ProtocolConformance *conformance) {
   auto module = conformance->getDeclContext()->getParentModule();
+  if (!CurGenericSignature && conformance->getGenericSignature()) {
+    CurGenericSignature =
```
Comment (DougGregor) [inline lib/AST/ASTMangler.cpp]: Hmm. I think this is correct, but if we set `CurGenericSignature` conditionally like this, we should remove it as we exit the function.

**Review by DougGregor — CHANGES_REQUESTED** [review]

File: lib/AST/ASTMangler.cpp (hunk 1 above)
Comment (theblixguy) [reply lib/AST/ASTMangler.cpp]: @DougGregor Sure, I have used `llvm::SaveAndRestore` to do that.

**Comment by jckarter** [comment]
@swift-ci Please test

**Review by DougGregor — APPROVED** [review]
Thank you!

**Pull request merged**